synapse/storage/databases/main/devices.py

```diff
@@ -57,6 +57,37 @@ def __init__(self, database: DatabasePool, db_conn, hs):
                 self._prune_old_outbound_device_pokes, 60 * 60 * 1000
             )
 
+    async def count_devices_by_users(self, user_ids: Optional[List[str]] = None) -> int:
+        """Retrieve number of all devices of given users.
+        Only returns number of devices that are not marked as hidden.
+
+        Args:
+            user_ids: The IDs of the users which owns devices
+        Returns:
+            Number of devices of this users.
+        """
+
+        def count_devices_by_users_txn(txn, user_ids):
+            sql = """
+                SELECT count(*)
+                FROM devices
+                WHERE
+                    hidden = '0' AND
+                    user_id IN ({})
+            """.format(
+                ",".join("?" for _ in user_ids)
+            )
```

Comment:
There's a helper function called `make_in_list_sql_clause` that makes this easier, e,g.:

https://github.com/matrix-org/synapse/blob/df4b1e9c74d56d79c274149b0dfb0fd5305c7659/synapse/storage/databases/main/events_worker.py#L920-L924